### ISSUE TITLE
Crash when internet is offline, but router is connected.

### ISSUE THREAD

**Issue by DjBlades** [libretro/RetroArch#4578]
## Description
I have a fairly unstable internet connection. At times, my internet goes down entirely, resulting in my router endlessly trying to connect for long periods, and my PC sitting in the "Connected, No internet access" state. During this state, Retroarch will crash immediately upon trying to load any content. The last few entries in the command prompt before it does references Cheevos, so I imagine it's related to RetroAchievements? If I disable my ethernet port entirely, or just remove the cable, Retroarch runs completely fine again. As expected, there are no problems when I have an active internet connection either. Only when my router is not able to connect to the internet.
![](http://i.imgur.com/p8Suwhc.png)

### Expected behavior

Content to load after being selected.

### Actual behavior

The command prompt freezes, eventually bringing up Windows' "Retroarch has stopped working" warning.

### Steps to reproduce the bug

1. Remove your router's connection to the internet. Disconnecting your ADSL cable is probably the best method to replicate it.
2. Try to run a game through Retroarch.
3. The command prompt will freeze, and Retroarch will crash.
4. Disconnect your computer from your router, either by removing the cable or just disabling your ports.
5. Try to run a game through Retroarch again.
6. Retroarch will now run completely fine.

### Version/Commit
You can find this information under Information/System Information

- RetroArch: 1.4.0/2c97e33

### Environment information

- OS: Windows 7 Home Premium

**Comment by retro-wertz** [comment]
i confirm this.

**Comment by fr500** [comment]
It should proceed after a few seconds. I use it w/o internet and it works fine.

**Comment by RobLoach** [comment]
> fail more gracefully

:+1:

**Comment by leiradel** [comment]
Until we implement a cache for the cheevos, making the request a task won't do any good since we'll have to wait for the result before starting the game. But yes, it has to fail with grace :)

**Comment by twinaphex** [comment]
For now, disable cheevos if it's truly causing you such trouble.

@leiradel You'll understand right that this is less, less than ideal right? The current implementation as-is? Please provide a proper solution because this situation can't stand as-is.

**Comment by leiradel** [comment]
Using a task doesn't automatically solve all the problems, it's still perfectly possible to make mistakes using it.

I'll see what is causing the crash and I'll try to fix it. My bets are all on net_http_get, which is not cheevos code.

**Comment by leiradel** [comment]
I can't repo this problem, the HTTP request just errors with

    RetroArch [ERROR] :: CHEEVOS error during HTTP GET: connect error.

and the game starts with no achievements loaded.

If someone can repo it with a debugger and paste a stack trace here it would help a lot.

**Comment by retro-wertz** [comment]
when Retroarch sometimes is not able to connect to online updater(task
failed something, i cannot remember the exact wordings) this is where it
just hangs when launching a game with retroarchievement as OP stated. a
"delay" in connection causing "RetroArch [ERROR] :: CHEEVOS error during
HTTP GET: connect error." is somewhat different since i mostly get this
while my internet is congested but still able to connect to "Online
Updater";

**Comment by leiradel** [comment]
I've connect to a router without an Internet connection and got that result, that's why I said that I couldn't repo the bug.

**Comment by DjBlades** [comment]
I assume for the debugger I just run the Retroarch_debug file. How can I get a stack trace?

**Comment by rootfather** [comment]
You have to use a debugger like "gdb" and run the retroarch_debug file (which contains all the necessary debug symbols) within the debugger.

**Comment by leiradel** [comment]
@DjBlades Do you have gdb available? If so:

```
$ gdb retroarch_debug
(gdb) run
<repo the issue, the crash will pause execution>
(gdb) bt
<stack trace is printed>
(gdb) quit
```

**Comment by retro-wertz** [comment]
this is more of a connection problem(delay reply from retroachievement?),
if you wait a bit it should just continue but no achievements since there
was delay or no response.

**Comment by retro-wertz** [comment]
this is really really hard to duplicate this for troubleshooting but is something that is happening for me too. ill try to explain as how i am experiencing this:

-as of the moment, retroarch cannot connect to server(or buildbot) when using Online Updater, throws a "Task Failed: Downloading  '.index-extended'. i have internet and connection is fine but retroarch cannot connect to Online updater. now with Achievements ON in this condition, running a game will either cause that one in OP screenshot, or just closes retroarch immediately with the last line in log shows cheevos initializing. it does not matter what game you are loading, even Dinothawr that has no achievements crash.

ra_log - https://hastebin.com/ebugekabal.sql

the only way to be able to play a game with this condition is to turn OFF Achievements. i do not know how internet traffic is being transfered as disconnecting my network does not recreate the scenario.

https://hastebin.com/ebugekabal.sql

**Comment by DjBlades** [comment]
I'm having trouble using gdb to debug Retroarch. I used Mingw to download it, but trying to start it at all just gives an error that I'm missing "libgcc_s_dw2-1.dll".

**Comment by leiradel** [comment]
@DjBlades did you install msys2?

**Comment by fr500** [comment]
@DjBlades it seems you're missing the redist, just get retroarch.7z

**Comment by DjBlades** [comment]
I downloaded msys2, exact same error.

**Comment by retro-wertz** [comment]
managed to get into the same(or similar) situation again, and this time was able to get backtrace:

https://hastebin.com/odicotirek.pl

hope this helps find whatever is causing this to crash.

**Comment by leiradel** [comment]
It seems net_http_get can return NET_HTTP_GET_OK but a NULL body. Cheevos code was considering NET_HTTP_GET_OK as the only success condition, and was crashing a little later trying to use the invalid body. I've changed it to also check the body.

Fixed at https://github.com/libretro/RetroArch/pull/4648

**Comment by retro-wertz** [comment]
ok commit done for these change. awaiting for the right situation to come
to test this out, or probably OP can compile based on latest commit and
test it, Online Updater should give Task Failed: for the test

**Comment by meleu** [comment]
@retro-wertz @DjBlades , can you try to reproduce it again with an updated version of RetroArch? I think it's not happening anymore.

**Comment by DjBlades** [comment]
Oh wow, I completely forgot about this.
Just tested now and everything is working completely fine, sorry for taking so long to update.
